Update feature-nrf52-sdk15 for v5.12

TacoGrandeTX opened the pull request:
### Description

Keep feature-nrf52-sdk15 current with master for v5.12 release:

* Rebase with master
* Added support for Arm Compiler 6 (AC6)
* Removed SDK14 files (which were never required)

Do not merge - this is a feature rebase only

### Pull request type
    [ ] Fix
    [ ] Refactor
    [x] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers


### Release Notes
Successful greentea results attached. There are two failures for nRF52840_DK with AC6 but those tests pass with gcc so appears to be a GT problem with these tests on nRF52840 under AC6. There is one failure on nRF52_DK related to TRNG as that always times out on laptop (as before). @desmond-blue conducted BLE testing using standard BLE examples.

[v5.12_tests.zip](https://github.com/ARMmbed/mbed-os/files/2943397/v5.12_tests.zip)

TacoGrandeTX committed:
Baseline legacy build using 14.2 drivers.

TacoGrandeTX committed:
Mods for baseline 15.0 support

TacoGrandeTX committed:
Bringing in Softdevice but excluding BLE feature and improvements in feature defines

TacoGrandeTX committed:
First porting effort covered by JIRA IOTDEV-1572:
- top level files ported from TARGET_NORDIC/TARGET_NRF5x/

Also addressed:
- fixed linking issue for gcc
- added support for nRF52-DK builds, but reverted to using nRF52840 sdk_config.h (must be updated)
- introduced "RTC" to targets.json (might need to be removed eventually)

TacoGrandeTX committed:
Ported analogin and pwm_out for IOTDEV-1573.

Also addressed:
- removed dependency on legacy config (excluded apply_old_config.h)
- removed legacy pwm and saadc headers
- Arm Compiler 5 linking issue (a band-aid for now... needs to
  be properly addressed for peripheral sharing)
- added missing header in SoftDevice file

Commit:
NRF52: i2c_api: port from Softdevice 14.2 to 15.0

Commit:
NRF52: spi_api: port from Softdevice 14.2 to 15.0

Commit:
serial_api: add changes to port from Softdevice 14.2 to 15.0

Commit:
serial_api: Fixes to compile with SDK15.0

TacoGrandeTX committed:
Updates to PWM for IOTDEV-1573

- Missed some NRFX defines that needed to changed
- Set PWM base clock to 125kHz (needs to be reverted back to 1 MHz)
- Updated sdk_config.h for nRF52_DK builds
- Brought in updates from PR7779 (fix for nRF52 PWM issues)

Commit:
Nordic sdk_config: Configure UART instances

Enable UART interfaces to enable serial debug prints.

Commit:
serial_api: Fix the interrupt registration for Nordic Serial Device

Using the nrfx_get_irq_number only works with the handle. Since we
know the IRQ numbers for UART0, RTC2 and EGU0, use them directly.

TacoGrandeTX committed:
Updates for IOTDEV-1575 and nRF52_DK builds
- Use new NRFX header file
- Fix nRF52832 linker script/ld files (hardcoded addresses for now)
- Temporarily remove DEVICE_TRNG for nRF52832 (which broke the build)
- Improve serial_putc() fix so we don't rely on "extra" functions

Commit:
NRF52: Fix vector table

Ensure that vector table gets initialized properly. The table that we
initialize in startup_nrf52840.S gets wiped out as the section is
declared as noinit. Fix this by implementing the weak function mbed_sdk_init
that inits the vector table.

Commit:
NRF52832: enable UART

Commit:
NRF: enable PRS (Peripheral Resource Sharing) for NRF52832 and NRF52840

TacoGrandeTX committed:
Updates for IOTDEV-1576 (only TRNG API)
- Added legacy nrf_drv_rng.c as there is no merit in fully adopting nrfx_rng.c
- Added nrf_queue library component
- Removed apply_old_config.h (unused for some time now)
- Updated sdk_config.h for queue and RNG support for nRF52832
- Brought back RNG into targets.json for nRF52832
Note: nRF52840 still uses CryptoCell 310 for TRNG

TacoGrandeTX committed:
nRF52832 needs NRFX_RNG_ENABLED to use the legacy driver

TacoGrandeTX committed:
nRF52_DK: Use mbed_lib.json to set app_offset (MBED_APP_START)

When the SoftDevice (SD) is enabled we need to set the app_offset
to 0x26000 to make room for the SoftDevice.  If we let the build
tools self-manage this, MBED_APP_START is coming out at 0x25000
which is not correct for the Nordic 15.0 SDK.

The app_offset value is translated to MBED_APP_START by the build
infrastructure.  We were hard coding MBED_APP_START in the scatter
and ld files as a temporary measure while testing.  This now sets
it properly if the SD is being brought in.

TacoGrandeTX committed:
Updates after rebase with Mbed master (circa OS v5.10.1)

* Remove NRF_SDH_CLOCK parameters from mbed_lib.json
* Bring in QSPI for nRF52840
* Migrate legacy QSPI driver to SDK v15 (nrfx_qspi.h)
* Remove outdated comment in i2c_api.c

TacoGrandeTX committed:
Fix build for nRF51 (fewer common files with nRF52)

The 15.0 SDK doesn't support the nRF51 so it must continue to use the legacy
NRF drivers.  Thus the original common_rtc.c, gpio_api.c and us_ticker.c are
restored and placed under TARGET_NRF5x/TARGET_NRF51.

Likewise the modified common_rtc.c, gpio_api.c and us_ticker.c are moved to
TARGET_NRF5x/TARGET_NRF52 so they are excluded from nRF51 builds.

TacoGrandeTX committed:
Remove dead #define PORT_SHIFT

nRF5x PinNames.h never utilized PORT_SHIFT so removed for clarity after
user commented on it.

Commit:
Add missing SoftDevice build support (MBR, NONE, OTA)

- Add MBR, NONE and OTA SoftDevice build options for S132 and S140
- Add S112 SoftDevice (single build option)
- Some folder restructuring in TARGET_SOFTDEVICE_COMMON was required

TacoGrandeTX committed:
Revert "NRF52: Fix vector table"

This reverts commit 3d2fa535a71b7bad08c6a91a31d74f4492c8948c.

This was a breaking change for the "MBR" and "NONE" builds.
After testing it was also determined that copying the vector
table a second time wasn't required for the "FULL" build.

TacoGrandeTX committed:
Cleanup for feature release

* Update TARGET_NRF5x/README.md to improve "Changing SoftDevice" section
  and added section on NRF52840 CryptoCell310 Support
* Update the file list in TARGET_SDK_15_0/TARGET_SOFTDEVICE_COMMON/README.md
* Add missing CR-LF to Nordic-provided SDK file
* Rename a header file in the TARGET_SOFTDEVICE_S112 tree

TacoGrandeTX committed:
Enabling NRFX_PPI_ENABLED for serial_api updates

TacoGrandeTX committed:
Changes required for rebase:
 - Brought in new nrfx APIs
 - Brought in PPI additions
 - Removed dead code for RTC

TacoGrandeTX committed:
Set the PWM clock prescaler back to default (1MHz clock)

Commit:
NRF52: serial: fix build warnings

Commit:
NRF52: Cleanup unused code

desmond-blue committed:
Update BLE SoftDevice from SDK14 to SDK15

    - Adjust memory for SoftDevice
    - Enable PRIO=5 for interrupt priority check
    - Change NRF_SD_BLE_API_VERSION to 6
    - Add handle and buffer for advertising and scanning
    - Remove guard for phy update
    - Change scatter files and mbed_lib.json for PR #8607

desmond-blue committed:
Fix timer1 irq configuration

desmond-blue committed:
Fix SoftDevice configuration on NRF52_DK

TacoGrandeTX committed:
Remove duplicate file

TacoGrandeTX committed:
Remove RTC from nRF52840_DK and DELTA_DFCM_NNN50 (was testing)

TacoGrandeTX committed:
Update file header at TARGET_NRF52 level
 - Update (c) year and company name
 - Add SPDX-License-Identifier
 - Impacted files:
    - Files touched for SDK15 port
    - Files added in 2018

desmond-blue committed:
Add BLE feature to mbed_lib.json of SoftDevice

Commit:
NRF52: reloc_vector_table: check the RTS pin settings in code

Move the check from preprocessor to the if condition

desmond-blue committed:
Fix nrf section iter macro check

TacoGrandeTX committed:
Rebase to master in preparation for v5.12.0

TacoGrandeTX committed:
Update QSPI in the SDK to support long frame mode

* Carry forward changes made to v14.2SDK to support long frame mode
(from commit 2a72bb67a315fb0e3a6654bd8ede27c4f579fa93).
* Fixed a few driver calls from rebase

TacoGrandeTX committed:
Fix SOFTDEVICE_NONE builds and clean up

* Update all scatterfiles to define MBED_RAM_START and MBED_RAM_SIZE.  These
get defined in mbed_lib.json when building with a SoftDevice but were missing
when no SoftDevice is brought in.
* Remove SDK 14.2 SoftDevice files.
* Add missing header file to qspi_api.c.

TacoGrandeTX committed:
Fix preprocessor conditionals for AC6 build

TacoGrandeTX committed:
Fix dangling #endif from rebase

paul-szczepanek-arm commented on targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_NRF52/spi_api.c:
```diff
@@ -44,21 +44,27 @@
 #include "pinmap_ex.h"
 #include "PeripheralPins.h"
 
-#include "nrf_drv_spi.h"
```
does that break old SDKs?

TacoGrandeTX replied:
The new header ```nrfx_spi.h``` replaces ```nrf_drv_spi.h```.  Ultimately though there is no backward compatibility in our move to SDK 15.  The Nordic SDK includes a legacy folder (https://github.com/TacoGrandeTX/mbed-os/tree/denmark/targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_SDK_15_0/integration/nrfx/legacy) but we've only carried forward the few files that are required for Mbed.    

I've just reviewed the files there and most of them can go away, so thanks for bringing this up.

TacoGrandeTX replied:
@paul-szczepanek-arm I've just removed the unnecessary legacy files with https://github.com/TacoGrandeTX/mbed-os/commit/833ed6392892b924ec9d3d6b8f7aae0a43746b56

TacoGrandeTX committed:
Remove redundant legacy files and definition of PACKED

* The following v14.2 files are removed from nrfx/legacy:
  nrf_drv_clock.h
  nrf_drv_common.h
  nrf_drv_gpiote.h
  nrf_drv_spi.h
  nrf_drv_twi.h
  nrf_drv_uarth.h
* Remove unneeded references to "nrfx_glue.h" and "nrf_drv_common.h"
* Remove the definition of PACKED from app_util_platform.h (already defined
  in mbed_toolchain.h)

TacoGrandeTX committed:
Fix targets.json for MTB_LAIRD_BL654

When removing CryptoCell310 we now need to define:
  NRFX_RNG_ENABLED=1
  RNG_ENABLED=1
  NRF_QUEUE_ENABLED=1
These are brought in via "macros_add".
